### DISCUSSION TITLE
Run on Android device

### DISCUSSION THREAD

**Discussion by ricfor** [dvmarinoff/Auuki#135] [General]
Is it possible to run Flux into an Android device as a native app ?
Where I can find resource to learn PWA  in order to work with BLE and ANT+ devices ?  
For example I need to control oder devices or change some visualizations, add translations etc.
Congratulations for your excellent work.
I paired Flux with my Elite Suito trainer and Wahoo HRM and seems OK.
Tomorrow I want to try to pair Assioma Pedals and check a workout.
Best regards

**Comment by dvmarinoff** [comment]
Hi @ricfor, and welcome to the repo!

## Run on Android as a native app

You can’t really access native functionality from PWAs, they remain web apps, but with the ability to install on Home Screen and cashe resources for offline use. This ability is present in the current code, but it is disabled.

## Learn PWA
Assuming you already know HTML, CSS and Javascript, this is a good place to start: https://web.dev/learn/pwa/

## Work with BLE and ANT+ devices

Here are some resources for Web-bluetooth:

A guide: https://developer.chrome.com/articles/bluetooth/
Code examples: https://googlechrome.github.io/samples/web-bluetooth/
The Specification document: https://webbluetoothcg.github.io/web-bluetooth/

While Web-bluetooth is a standart thing, WebANT+ is just a personal project of mine. In 2021 Google shipped into Chrome Web Serial API, which along with the Web USB API, made communicating with ANT+ USB dongle from a web page possible. While my project has big chunks of the ANT+ protocol implemented it remains a prove of concept.

I am currently stuck on a bug, where the connection stops reading every 4 minutes, and then resumes after another 4 minutes, and keeps repeating. I am not systems engineer and writing low level USB code goes a bit over my head. I tried reading the USB spec, but that thing is pretty big. If you have any experience with this kind of stuff it would be great if you can help.

**Comment by ricfor** [comment]
Thank you Dimitar for your kind reply and for the valuable advice you gave me. 
Unfortunately I'm not a system engineer either and I have no experience writing USB code but if I can dig deeper I'll be happy to help you.
Today I did some small tests using the Flux version you shared on the web accessing it by an Android smartphone.
I succesful connected my Elite Suito Trainer in BLE and I also connect HRM Wahoo and the Assioma power meter pedals reading perfectly the data.
On next days I hope to find the time to try a complete training session. 
OFFLINE APP
I concentrated also on getting the app to work offline because normally my trainer is in a room where there isn't a good wifi connection.
I copied the PWA files in a windows PC and then also in a Raspberry on which I previously installed node.js and lite-server .
However, a strange behaviour happens: connecting to both the offline servers with my android smartphone, Flux starts correctly but I cannot connect the BLE devices and I get this error :
![Image 1](https://user-images.githubusercontent.com/23704248/202924063-cbbec72d-8b90-4c83-aa96-472b9ebc48e2.jpg)


Maybe you can help me to figure out where I'm wrong?

**Comment by dvmarinoff** [comment]
Web-bluetooth requires https, getting an error at requestDevice looks like this could be the issue.

If you decide to make the app is installable on your fork, which will also get you offline use, here are the steps:

1. Uncomment this line in `index.js`:
https://github.com/dvmarinoff/Flux/blob/a5f956763f43e1d0355cc00aa36026846cfaa845/src/index.js#L25

2. Make sure the file list in `sw.js` is up to date. The browser console will guide you with the errors until it is right.
https://github.com/dvmarinoff/Flux/blob/a5f956763f43e1d0355cc00aa36026846cfaa845/src/sw.js#L2

**Reply by ricfor** [reply]
Thank you, the advice you gave, allowed me to run the app on the local server.
I also found the time to carry out a training session by pairing the app  with the Elite Suito trainer and the other bluetooth devices and everything worked really well.
I'd like to make a small change to the local app to be able to increase the resistance in manual mode by 1 units instead of 10, because it seems that increasing by 10 doesn't work well with the Suito

**Comment by dvmarinoff** [comment]
@ricfor 

Yeah, sure. You will need to add `, step: 1` to the following line: 
https://github.com/dvmarinoff/Flux/blob/a5f956763f43e1d0355cc00aa36026846cfaa845/src/models/models.js#L1007

so that it becomes:

``` 
const resistaceTarget = new resistanceTarget({prop: 'resistanceTarget', step: 1}) 
```

**Reply by ricfor** [reply]
I have to congratulate you because the application works really well.
Following your instructions I modified the resistance increment steps and they are now ideal for my trainer.
I also managed to install the app on a local server on Raspberry, thanks to which I can connect with the smartphone, when I'am in my garage where there is no internet connection.
I uploaded some personal workouts and the trainer regulation works very well
In my opinion, a possible improvement would be to enlarge the command buttons because on a smartphone they are quite small and when you are under stress it becomes difficult to be precise with the click.
Another improvement could be to insert a button that allows to move to the next step of a workout.
This becomes useful for example during a hiit workout with many all out steps, in the case you reach exhaustion before having completed all the planned steps and you want to jump quickly to the final recovery phase.
Another improvement, which I think you have already been advised, could be to have the possibility to save your workouts. However, I have noticed that it is quite easy to customize the workouts.js file
Thank you very much for your work and for your advice, also because I'm slowly discovering the PWAs world, that I didn't know before

**Reply by dvmarinoff** [reply]
Hey, thanks!

> enlarge the command buttons

The player buttons I guess can get a little bigger. Will see how much would be possible.

> button that allows to move to the next step of a workout

The Lap button would move to the next interval during a structured workout, and would create new Lap during free-ride. That's the one with the Stop Watch.

> save your workouts

Just working on that right now. I am planning to use Intervals.icu as a backend. But still figuring out how to do OAuth in a PWA. It’s going to be something like this eventually: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps#section-6.4.2.

Right now I got a local branch that syncs Workout of the Day and stores completed Activities to Intervals.icu using http basic authentication. But OAuth is better approach.
